### ISSUE TITLE
Custom icon iconSize Inconstancies

### ISSUE THREAD

**Issue by HPieters** [react-native-mapbox-gl/maps#68]
**Describe the bug**
Custom icons have inconsistencies between debug and release. In debug mode the icons appear roughly 3 time as small as they are in release. This is a bug that has been in the library since 6.1.3 and has been reported twice before in [https://github.com/nitaliano/react-native-mapbox-gl/issues/1348](https://github.com/nitaliano/react-native-mapbox-gl/issues/1348) and [https://github.com/nitaliano/react-native-mapbox-gl/issues/1417](https://github.com/nitaliano/react-native-mapbox-gl/issues/1417)

**To Reproduce**
I created a [example project that has this issue on github](https://github.com/HPieters/mapbox-icon-size-issue). To reproduce create a new project and copy the content from App.js in the repo to your project. 

**Expected behavior**
No difference in icon size or scaling between debug and release builds.

**Screenshots**
Running [issue example project from github](https://github.com/HPieters/mapbox-icon-size-issue).

_Debug Scheme_
![debug](https://user-images.githubusercontent.com/2362116/58547959-b0426000-8208-11e9-8489-962d329ece24.png)

_Release Scheme_
![release](https://user-images.githubusercontent.com/2362116/58547968-b33d5080-8208-11e9-8ab3-b0d03848f066.png)

For more screenshots see [https://github.com/nitaliano/react-native-mapbox-gl/issues/1417](https://github.com/nitaliano/react-native-mapbox-gl/issues/1417)

**Versions (please complete the following information):**
 - Platfrom: iOS (But I think the problem also exists on Android)
 - Device: Simulator iPhone Xr / iPhone 7 
 - OS: Latest (12.2)
 - SDK Version Latest
 - React Native Version: Latest (0.59.x)

**Additional context**
As mentioned this problem has been reported before and seems to have been introduced in version 6.1.3. In 6.1.2 I didn't have these problems. [This commit](https://github.com/nitaliano/react-native-mapbox-gl/commit/848604edf29bda4596df5255cddc1ee6784c356f) has been added in 6.1.3 and might be the cause, not sure though.

**Comment by mfazekas** [comment]
I was able to reproduce the issue seems to be that the same .png UIImage when loaded from `file://` url (production) and when loaded from `http` url has different scale.

It seems to be that in file case ReactNative uses different code path to load than in the url case.
Maybe different RN versions has different issues.

**Comment by mfazekas** [comment]
It looks like the 6.1.3 solution works fine when multiresolution images are provided with `image@3x.png`. And 6.1.2 does not work for them. But for regular images 6.1.2 works fine - and 6.1.3 is broken. 

On 6.1.2 in dev bundle mode (http) `img@3x.png` is rendered as 3x size as scale remains 1.
On 6.1.3 in dev bundle mode (http) `img.png` is renderd as 1/3x the size as scale set to 3.

**Comment by HPieters** [comment]
@mfazekas Thanks for taking the time and taking a look at this issue! Once I have a bit off time I will test the 3x.png "workaround" and see if it solves the issue in my apps.

**Comment by HPieters** [comment]
@mfazekas Thanks once again, seems like your fix worked for iOS however it does cause an error for me on Android version:

```
06-03 15:41:54.188  1856  1900 E AndroidRuntime: FATAL EXCEPTION: AsyncTask #3
06-03 15:41:54.188  1856  1900 E AndroidRuntime: Process: com.demo.app, PID: 1856
06-03 15:41:54.188  1856  1900 E AndroidRuntime: java.lang.RuntimeException: An error occurred while executing doInBackground()
06-03 15:41:54.188  1856  1900 E AndroidRuntime: 	at android.os.AsyncTask$3.done(AsyncTask.java:354)
06-03 15:41:54.188  1856  1900 E AndroidRuntime: 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
06-03 15:41:54.188  1856  1900 E AndroidRuntime: 	at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
06-03 15:41:54.188  1856  1900 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:271)
06-03 15:41:54.188  1856  1900 E AndroidRuntime: 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
06-03 15:41:54.188  1856  1900 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
06-03 15:41:54.188  1856  1900 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
06-03 15:41:54.188  1856  1900 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:764)
06-03 15:41:54.188  1856  1900 E AndroidRuntime: Caused by: java.lang.ClassCastException: java.lang.String cannot be cast to com.mapbox.rctmgl.utils.ImageEntry
```

**Comment by mfazekas** [comment]
@HPieters sorry yep it seems Java is not as typesafe as i hoped to be:

This one needs to be mapped to the new structure:
https://github.com/react-native-mapbox-gl/maps/blob/8835e678c1a1d1232e9c47c7fb00c89377dcabc7/android/rctmgl/src/main/java/com/mapbox/rctmgl/components/styles/sources/RCTMGLShapeSource.java#L77-L79


Like this one:
https://github.com/react-native-mapbox-gl/maps/blob/f6439672da9daa2bec4eae027caecd31e58d702e/android/rctmgl/src/main/java/com/mapbox/rctmgl/components/styles/RCTMGLStyle.java#L78-L81

**Comment by jonong** [comment]
@mfazekas it appears that the deprecation of `ShapeSource#images` and the implementation of the new `Images` component has caused the images to appear twice or thrice as large, depending on the screen scale.

**Comment by mfazekas** [comment]
@jonong can you post a concrete example and exact environment - there are many ways to specify an image (url, asset on prod, asset on dev, path, native asset), also is it android or ios, not sure which one you're using. Also does it works with `ShapeSource#images` but not with `Images#images` - note you can still use `ShapeSource#image` it just gives a deprecation warning.

**Comment by jonong** [comment]
Hey @mfazekas sorry for the late reply. Here's a sample project: https://github.com/jonong/MapboxTest

I'm running React Native 0.59.9 with Mapbox 7.0.0-rc3. I've tested the sample project on iOS, and still experience the image scale issue.

**Comment by mattijsf** [comment]
I used to work on something similar on the old repo https://github.com/nitaliano/react-native-mapbox-gl/pull/1206

Not sure yet what is happening now but I also noticed that images are not rendered the size that I expected.

In the old repo images were renders at the same size on the map as if you would see them when provided to an `<Image />` view component. That's currently no longer the case. I loved to dive into this some day but I'm not sure if I can. As a workaround I currently use `1 / PixelRatio.get()` for local images. I haven't tried prod builds yet, might run into issues there if I see this issue.

**Comment by mfazekas** [comment]
@mattijsf ideally the code in the new repo is improvement over the old one.

The issue with nitaliano/react-native-mapbox-gl#1206 was that it assumed that image has native scaling. If the image did not have native scaling (like your device was 3x, but image had no `@3x.png` variant, then the old code still asumed 3.0 scale, but the images were not 3x scale, so they rended smaller.

The new code now uses the scale attribute that's generated by require/import 'image' statements, instead of assuming the native device scale. That should work for both the case where you have native images, eg `@3x.png`  variant for 3x device and where you doesn't like 3x device but no variants.

**Comment by jonong** [comment]
@mattijsf I use `1 / PixelRatio.get()` as an interim solution as well. However, I've noticed that this does not work well when using a `SymbolLayer` with `symbolPlacement` set to `line` and a defined `symbolSpacing`. It appears that the symbol positions are calculated based on a fixed `iconSize` of `1`, and does not update when I reduce the `iconSize` using the above method.

**Comment by mfazekas** [comment]
@jonong, @mattijsf  i've found a bug in code which should be fixed by #191 let me know if you still see an issue with image sizing.

**Comment by jonong** [comment]
@mfazekas I've tested with the latest commit and it works great! Thanks!
